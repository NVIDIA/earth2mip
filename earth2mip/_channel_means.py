# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

channel_means = {
    "u10m": -0.06854962557554245,
    "v10m": 0.18743863701820374,
    "u100m": -0.009113931097090244,
    "v100m": 0.19518691301345825,
    "t2m": 278.7031555175781,
    "sp": 96653.515625,
    "msl": 100957.6796875,
    "tcwv": 18.46604347229004,
    "u1": 10.748488426208496,
    "u2": 9.040291786193848,
    "u3": 7.640656471252441,
    "u5": 6.190047264099121,
    "u7": 5.613990306854248,
    "u10": 4.81683874130249,
    "u20": 3.7416577339172363,
    "u30": 3.9482882022857666,
    "u50": 5.662464618682861,
    "u70": 7.420891761779785,
    "u100": 10.338683128356934,
    "u125": 12.378543853759766,
    "u150": 13.560766220092773,
    "u175": 14.09630012512207,
    "u200": 14.167032241821289,
    "u225": 13.864096641540527,
    "u250": 13.288179397583008,
    "u300": 11.738770484924316,
    "u350": 10.155152320861816,
    "u400": 8.77362060546875,
    "u450": 7.578890800476074,
    "u500": 6.538493633270264,
    "u550": 5.618127822875977,
    "u600": 4.793304920196533,
    "u650": 4.032490253448486,
    "u700": 3.3098866939544678,
    "u750": 2.626377820968628,
    "u775": 2.299865484237671,
    "u800": 1.9830362796783447,
    "u825": 1.6734198331832886,
    "u850": 1.3726757764816284,
    "u875": 1.0908571481704712,
    "u900": 0.8307265639305115,
    "u925": 0.5900797843933105,
    "u950": 0.37076279520988464,
    "u975": 0.17277194559574127,
    "u1000": -0.05516379326581955,
    "v1": -0.081306092441082,
    "v2": -0.021461497992277145,
    "v3": -0.020173462107777596,
    "v5": -0.00731061352416873,
    "v7": -0.004137726966291666,
    "v10": -0.0019085388630628586,
    "v20": -0.00026605380116961896,
    "v30": 0.0010730322683230042,
    "v50": 0.006631731055676937,
    "v70": 0.0045732553116977215,
    "v100": 0.0040822383016347885,
    "v125": -0.02326071448624134,
    "v150": -0.045998428016901016,
    "v175": -0.0552876852452755,
    "v200": -0.03847499191761017,
    "v225": -0.02821849100291729,
    "v250": -0.027320099994540215,
    "v300": -0.02411925420165062,
    "v350": -0.01592441089451313,
    "v400": -0.019759656861424446,
    "v450": -0.026930494233965874,
    "v500": -0.02907356433570385,
    "v550": -0.031823400408029556,
    "v600": -0.02572263590991497,
    "v650": -0.009626002050936222,
    "v700": 0.02697981894016266,
    "v750": 0.06845545023679733,
    "v775": 0.0888051986694336,
    "v800": 0.10753155499696732,
    "v825": 0.12397794425487518,
    "v850": 0.14166374504566193,
    "v875": 0.1662677526473999,
    "v900": 0.19087384641170502,
    "v925": 0.20176121592521667,
    "v950": 0.19572316110134125,
    "v975": 0.18371663987636566,
    "v1000": 0.18469774723052979,
    "w1": -1.1315158872093889e-06,
    "w2": -8.018057087610941e-06,
    "w3": -1.537994648970198e-05,
    "w5": -2.478267197147943e-05,
    "w7": -3.223319072276354e-05,
    "w10": -4.135979543207213e-05,
    "w20": -5.539332050830126e-05,
    "w30": -5.9556386986514553e-05,
    "w50": -5.383044117479585e-05,
    "w70": -4.982736209058203e-05,
    "w100": -2.3971153495949693e-05,
    "w125": 3.87470163332182e-06,
    "w150": 1.66643312695669e-05,
    "w175": 1.5288433132809587e-05,
    "w200": 9.045768820215017e-06,
    "w225": 6.168767413328169e-06,
    "w250": 1.4241597455111332e-05,
    "w300": 9.39343444770202e-05,
    "w350": 0.00019888172391802073,
    "w400": 0.000272307894192636,
    "w450": 0.0002964134037028998,
    "w500": 0.00030192555277608335,
    "w550": 0.0002988489577546716,
    "w600": 0.0003334358334541321,
    "w650": 0.0006918106810189784,
    "w700": 0.002709059976041317,
    "w750": 0.005279306787997484,
    "w775": 0.006850168574601412,
    "w800": 0.008400746621191502,
    "w825": 0.010028586722910404,
    "w850": 0.011793135665357113,
    "w875": 0.01368639525026083,
    "w900": 0.015604125335812569,
    "w925": 0.01744643785059452,
    "w950": 0.019302070140838623,
    "w975": 0.020728887990117073,
    "w1000": 0.020804615691304207,
    "z1": 463206.4375,
    "z2": 411553.125,
    "z3": 382229.46875,
    "z5": 346674.875,
    "z7": 324015.15625,
    "z10": 300542.71875,
    "z20": 256143.234375,
    "z30": 230781.734375,
    "z50": 199379.265625,
    "z70": 179031.015625,
    "z100": 157709.578125,
    "z125": 144308.1875,
    "z150": 133217.546875,
    "z175": 123719.1875,
    "z200": 115399.328125,
    "z225": 107985.0703125,
    "z250": 101282.53125,
    "z300": 89464.4921875,
    "z350": 79181.1875,
    "z400": 70015.5625,
    "z450": 61723.00390625,
    "z500": 54140.7421875,
    "z550": 47151.71484375,
    "z600": 40667.58984375,
    "z650": 34617.66796875,
    "z700": 28943.169921875,
    "z750": 23597.029296875,
    "z775": 21036.0859375,
    "z800": 18544.6875,
    "z825": 16119.4697265625,
    "z850": 13757.439453125,
    "z875": 11455.5458984375,
    "z900": 9210.625,
    "z925": 7019.46533203125,
    "z950": 4878.94189453125,
    "z975": 2786.269775390625,
    "z1000": 738.6001586914062,
    "t1": 261.1324462890625,
    "t2": 255.74183654785156,
    "t3": 247.80210876464844,
    "t5": 237.51869201660156,
    "t7": 231.79122924804688,
    "t10": 226.91737365722656,
    "t20": 219.69107055664062,
    "t30": 216.14366149902344,
    "t50": 212.13992309570312,
    "t70": 209.16539001464844,
    "t100": 208.21510314941406,
    "t125": 210.52633666992188,
    "t150": 213.35560607910156,
    "t175": 215.93296813964844,
    "t200": 218.19093322753906,
    "t225": 220.4385986328125,
    "t250": 222.94039916992188,
    "t300": 229.04507446289062,
    "t350": 235.85696411132812,
    "t400": 242.3114013671875,
    "t450": 248.05335998535156,
    "t500": 253.07688903808594,
    "t550": 257.4590759277344,
    "t600": 261.2414855957031,
    "t650": 264.5358581542969,
    "t700": 267.5036926269531,
    "t750": 270.2417297363281,
    "t775": 271.4933776855469,
    "t800": 272.6558532714844,
    "t825": 273.7077331542969,
    "t850": 274.6662902832031,
    "t875": 275.5856018066406,
    "t900": 276.5234680175781,
    "t925": 277.5076599121094,
    "t950": 278.5635681152344,
    "t975": 279.78955078125,
    "t1000": 281.2162170410156,
    "r1": 0.0012648305855691433,
    "r2": 0.006784914992749691,
    "r3": 0.021754322573542595,
    "r5": 0.10752345621585846,
    "r7": 0.2999992072582245,
    "r10": 0.8303617835044861,
    "r20": 3.4991626739501953,
    "r30": 4.6645050048828125,
    "r50": 6.580825328826904,
    "r70": 13.205595970153809,
    "r100": 26.424592971801758,
    "r125": 26.096942901611328,
    "r150": 26.452449798583984,
    "r175": 29.685396194458008,
    "r200": 35.18281936645508,
    "r225": 41.35318374633789,
    "r250": 46.84218978881836,
    "r300": 53.30179214477539,
    "r350": 53.954071044921875,
    "r400": 52.459781646728516,
    "r450": 51.048095703125,
    "r500": 50.507022857666016,
    "r550": 50.88429260253906,
    "r600": 51.785911560058594,
    "r650": 53.15390396118164,
    "r700": 55.32671356201172,
    "r750": 58.33307647705078,
    "r775": 60.228816986083984,
    "r800": 62.532310485839844,
    "r825": 65.50992584228516,
    "r850": 69.1156234741211,
    "r875": 72.80475616455078,
    "r900": 76.10183715820312,
    "r925": 78.84436798095703,
    "r950": 80.54322052001953,
    "r975": 79.97803497314453,
    "r1000": 77.97394561767578,
    "2d": 274.1190185546875,
    "sst": 189.70262145996094,
    "skt": 279.1072692871094,
    "stl1": 279.7412414550781,
    "swvl1": 0.08473234623670578,
}
