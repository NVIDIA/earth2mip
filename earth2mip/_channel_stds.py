# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

channel_stds = {
    "u10m": 5.582091331481934,
    "v10m": 4.789056301116943,
    "u100m": 7.004323482513428,
    "v100m": 6.123928546905518,
    "t2m": 21.191524505615234,
    "sp": 9595.267578125,
    "msl": 1344.2525634765625,
    "tcwv": 16.540952682495117,
    "u1": 38.58013916015625,
    "u2": 34.72814178466797,
    "u3": 32.2275390625,
    "u5": 29.21213722229004,
    "u7": 27.27058219909668,
    "u10": 25.552366256713867,
    "u20": 21.857524871826172,
    "u30": 19.176223754882812,
    "u50": 15.317777633666992,
    "u70": 13.463854789733887,
    "u100": 13.648713111877441,
    "u125": 14.856287002563477,
    "u150": 16.07061767578125,
    "u175": 17.02838897705078,
    "u200": 17.721553802490234,
    "u225": 18.051687240600586,
    "u250": 18.01706886291504,
    "u300": 17.13730239868164,
    "u350": 15.746345520019531,
    "u400": 14.337974548339844,
    "u450": 13.082659721374512,
    "u500": 12.004372596740723,
    "u550": 11.114582061767578,
    "u600": 10.387426376342773,
    "u650": 9.774086952209473,
    "u700": 9.236701965332031,
    "u750": 8.800454139709473,
    "u775": 8.619154930114746,
    "u800": 8.46338176727295,
    "u825": 8.338661193847656,
    "u850": 8.247283935546875,
    "u875": 8.179678916931152,
    "u900": 8.111303329467773,
    "u925": 7.978816032409668,
    "u950": 7.672853946685791,
    "u975": 7.052810192108154,
    "u1000": 6.180253028869629,
    "v1": 15.007347106933594,
    "v2": 14.02863883972168,
    "v3": 13.400466918945312,
    "v5": 12.586094856262207,
    "v7": 12.008319854736328,
    "v10": 11.252314567565918,
    "v20": 9.49392318725586,
    "v30": 8.383625984191895,
    "v50": 7.201364517211914,
    "v70": 6.877016067504883,
    "v100": 7.557221412658691,
    "v125": 8.542673110961914,
    "v150": 9.602500915527344,
    "v175": 10.747245788574219,
    "v200": 11.92225456237793,
    "v225": 12.88244342803955,
    "v250": 13.436037063598633,
    "v300": 13.366893768310547,
    "v350": 12.399846076965332,
    "v400": 11.22365951538086,
    "v450": 10.125510215759277,
    "v500": 9.184946060180664,
    "v550": 8.425360679626465,
    "v600": 7.8204240798950195,
    "v650": 7.334446430206299,
    "v700": 6.889240264892578,
    "v750": 6.559966087341309,
    "v775": 6.439513206481934,
    "v800": 6.349121570587158,
    "v825": 6.294491291046143,
    "v850": 6.284781455993652,
    "v875": 6.329617500305176,
    "v900": 6.4216156005859375,
    "v925": 6.501189708709717,
    "v950": 6.4423980712890625,
    "v975": 6.05946159362793,
    "v1000": 5.338335037231445,
    "w1": 0.0009700588998384774,
    "w2": 0.0017569169867783785,
    "w3": 0.002291186945512891,
    "w5": 0.0030673006549477577,
    "w7": 0.0038343495689332485,
    "w10": 0.004771752282977104,
    "w20": 0.007248147390782833,
    "w30": 0.0092012332752347,
    "w50": 0.012652132660150528,
    "w70": 0.016436472535133362,
    "w100": 0.02630731090903282,
    "w125": 0.038530394434928894,
    "w150": 0.05268237739801407,
    "w175": 0.06788258254528046,
    "w200": 0.08365459740161896,
    "w225": 0.09980383515357971,
    "w250": 0.11605729162693024,
    "w300": 0.1480696201324463,
    "w350": 0.1759636253118515,
    "w400": 0.19732540845870972,
    "w450": 0.2122214138507843,
    "w500": 0.22131618857383728,
    "w550": 0.22764532268047333,
    "w600": 0.23328272998332977,
    "w650": 0.23892800509929657,
    "w700": 0.24304194748401642,
    "w750": 0.24660667777061462,
    "w775": 0.24764472246170044,
    "w800": 0.24739933013916016,
    "w825": 0.24559083580970764,
    "w850": 0.24174177646636963,
    "w875": 0.23500598967075348,
    "w900": 0.223381906747818,
    "w925": 0.2050851732492447,
    "w950": 0.1813991665840149,
    "w975": 0.1576462835073471,
    "w1000": 0.14332318305969238,
    "z1": 17160.59375,
    "z2": 15051.7412109375,
    "z3": 13678.45703125,
    "z5": 12030.5791015625,
    "z7": 11000.9716796875,
    "z10": 9939.580078125,
    "z20": 7970.08447265625,
    "z30": 6934.6259765625,
    "z50": 5881.4306640625,
    "z70": 5497.146484375,
    "z100": 5545.66748046875,
    "z125": 5721.24609375,
    "z150": 5838.37841796875,
    "z175": 5871.626953125,
    "z200": 5826.142578125,
    "z225": 5712.2578125,
    "z250": 5541.61279296875,
    "z300": 5096.86474609375,
    "z350": 4614.15478515625,
    "z400": 4156.310546875,
    "z450": 3738.7666015625,
    "z500": 3361.076171875,
    "z550": 3018.40625,
    "z600": 2704.42138671875,
    "z650": 2413.234619140625,
    "z700": 2143.913818359375,
    "z750": 1898.7406005859375,
    "z775": 1784.596435546875,
    "z800": 1676.227783203125,
    "z825": 1573.9453125,
    "z850": 1478.46435546875,
    "z875": 1390.4481201171875,
    "z900": 1310.4091796875,
    "z925": 1238.84423828125,
    "z950": 1176.4459228515625,
    "z975": 1124.12060546875,
    "z1000": 1082.5850830078125,
    "t1": 12.217451095581055,
    "t2": 15.37381362915039,
    "t3": 15.199849128723145,
    "t5": 14.100061416625977,
    "t7": 13.47716236114502,
    "t10": 12.883852005004883,
    "t20": 11.544325828552246,
    "t30": 10.671643257141113,
    "t50": 10.337119102478027,
    "t70": 11.590261459350586,
    "t100": 12.401776313781738,
    "t125": 10.661434173583984,
    "t150": 8.748968124389648,
    "t175": 7.472810745239258,
    "t200": 7.142058849334717,
    "t225": 7.623969554901123,
    "t250": 8.593498229980469,
    "t300": 10.757156372070312,
    "t350": 12.078363418579102,
    "t400": 12.678500175476074,
    "t450": 12.939180374145508,
    "t500": 13.049148559570312,
    "t550": 13.15401840209961,
    "t600": 13.449394226074219,
    "t650": 14.161675453186035,
    "t700": 14.810254096984863,
    "t750": 15.129765510559082,
    "t775": 15.2506685256958,
    "t800": 15.36613655090332,
    "t825": 15.471822738647461,
    "t850": 15.559452056884766,
    "t875": 15.650764465332031,
    "t900": 15.786845207214355,
    "t925": 15.981817245483398,
    "t950": 16.2429141998291,
    "t975": 16.620704650878906,
    "t1000": 16.995594024658203,
    "r1": 0.01644953526556492,
    "r2": 0.045997656881809235,
    "r3": 0.0963272675871849,
    "r5": 0.38980916142463684,
    "r7": 1.0760670900344849,
    "r10": 3.1369779109954834,
    "r20": 13.323467254638672,
    "r30": 15.115065574645996,
    "r50": 15.014254570007324,
    "r70": 18.140913009643555,
    "r100": 33.699092864990234,
    "r125": 32.9305305480957,
    "r150": 32.013851165771484,
    "r175": 32.755245208740234,
    "r200": 34.16655731201172,
    "r225": 35.020626068115234,
    "r250": 34.8388671875,
    "r300": 34.21603775024414,
    "r350": 34.3671989440918,
    "r400": 34.230587005615234,
    "r450": 33.944515228271484,
    "r500": 33.60013198852539,
    "r550": 33.3435173034668,
    "r600": 32.77552795410156,
    "r650": 32.15131759643555,
    "r700": 31.51679229736328,
    "r750": 30.690650939941406,
    "r775": 30.05921745300293,
    "r800": 29.193891525268555,
    "r825": 28.011188507080078,
    "r850": 26.4312801361084,
    "r875": 24.70786476135254,
    "r900": 23.095003128051758,
    "r925": 21.721576690673828,
    "r950": 20.58991241455078,
    "r975": 19.374773025512695,
    "r1000": 18.462560653686523,
    "2d": 20.632549285888672,
    "sst": 136.16073608398438,
    "skt": 22.203195571899414,
    "stl1": 21.357025146484375,
    "swvl1": 0.14037616550922394,
}
