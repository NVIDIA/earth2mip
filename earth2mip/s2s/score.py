# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import altair as alt
import numpy as np
import pandas as pd
import xarray
import xskillscore


def score(terciles: xarray.Dataset, truth: xarray.Dataset) -> pd.DataFrame:
    """area-weighted RPSS scores

    Regions: global, northern hemisphere, and southern hemisphere

    Args:
        terciles: predicted terciles. must have `longitude`, `latitude`, `forecast_time` coordinates/dims.
        truth: true terciles. same format as terciles

    Returns:
        dataframe with scores for different regions and climatology. Example::

                lead_time forecast_time       t2m valid_time  week source         region        tp
            0        21.0    2018-01-02  1.455134 2018-01-23   1.0   sfno         Global       NaN
            1        35.0    2018-01-02  1.357457 2018-02-06   1.0   sfno         Global       NaN
            2        21.0    2018-01-02  1.308716 2018-01-23   NaN   clim         Global  1.310107
            3        35.0    2018-01-02  1.306281 2018-02-06   NaN   clim         Global  1.312259
            4        21.0    2018-01-02  1.331612 2018-01-23   1.0   sfno  Northern Hem.       NaN
            5        35.0    2018-01-02  1.211101 2018-02-06   1.0   sfno  Northern Hem.       NaN
            6        21.0    2018-01-02  1.184829 2018-01-23   NaN   clim  Northern Hem.  1.237482
            7        35.0    2018-01-02  1.180459 2018-02-06   NaN   clim  Northern Hem.  1.241959
            8        21.0    2018-01-02  1.575785 2018-01-23   1.0   sfno  Southern Hem.       NaN
            9        35.0    2018-01-02  1.497714 2018-02-06   1.0   sfno  Southern Hem.       NaN
            10       21.0    2018-01-02  1.431765 2018-01-23   NaN   clim  Southern Hem.  1.381933
            11       35.0    2018-01-02  1.430871 2018-02-06   NaN   clim  Southern Hem.  1.381818

    """  # noqa

    sfno = xarray.open_dataset(terciles)
    obs = xarray.open_dataset(truth)

    sfno, obs = xarray.align(sfno, obs)
    assert obs.sizes["forecast_time"] > 0  # noqa

    clim = xarray.ones_like(obs) / 3.0

    # %%
    masks = {}
    cos_lat = np.cos(np.deg2rad(obs.latitude))
    masks["Global"] = cos_lat
    masks["Northern Hem."] = cos_lat.where(obs.latitude > 0, 0.0)
    masks["Southern Hem."] = cos_lat.where(obs.latitude < 0, 0.0)

    scores = []
    for mask in masks:
        cos_lat = masks[mask]

        iscores = {}
        iscores["sfno"] = xskillscore.rps(
            sfno,
            obs,
            category_edges=None,
            input_distributions="p",
            dim=["latitude", "longitude"],
            weights=cos_lat,
        )
        iscores["clim"] = xskillscore.rps(
            clim,
            obs,
            category_edges=None,
            input_distributions="p",
            dim=["latitude", "longitude"],
            weights=cos_lat,
        )

        for key in iscores:
            v = iscores[key]
            scores.append(v.assign(source=key, region=mask))

    df = pd.concat([d.to_dataframe() for d in scores])
    df = df.reset_index()
    df["lead_time"] = df.lead_time + pd.Timedelta("7D")
    df["valid_time"] = df.lead_time + df.forecast_time
    df["lead_time"] = df.lead_time / pd.Timedelta("1D")
    return df


def plot(df, output, timeUnit="yearmonthdate"):
    import vl_convert

    alt.data_transformers.disable_max_rows()

    # https://davidmathlogic.com/colorblind/#%23000000-%23E69F00-%2356B4E9-%23009E73-%23F0E442-%230072B2-%23D55E00-%23CC79A7
    wong_palette = [
        "#000000",
        "#E69F00",
        "#56B4E9",
        "#009E73",
        "#F0E442",
        "#0072B2",
        "#D55E00",
        "#CC79A7",
    ]

    # altair

    def my_theme():
        return {"config": {"range": {"category": wong_palette}}}

    alt.themes.register("my_theme", my_theme)
    alt.themes.enable("my_theme")
    c = (
        alt.Chart(df)
        .mark_line()
        .encode(
            x=alt.X(field="valid_time", timeUnit=timeUnit),
            y=alt.Y("mean(t2m)", title="RPSS T2M"),
            column="region",
            color="source",
            row=alt.Row("lead_time", type="ordinal", title="Lead Time (days)"),
        )
        .properties(width=200, height=200 / 1.61)
    )

    png_data = vl_convert.vegalite_to_png(vl_spec=c.to_json(), scale=3)
    with open(output, "wb") as f:
        f.write(png_data)
