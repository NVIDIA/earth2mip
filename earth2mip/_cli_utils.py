# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import argparse
import datetime
from typing import List

from earth2mip import networks, schema


class TimeRange:
    @staticmethod
    def add_args(parser: argparse.ArgumentParser):
        parser.add_argument(
            "--start-time", default="2018-01-01", type=datetime.datetime.fromisoformat
        )
        parser.add_argument(
            "--time-step",
            default="12",
            type=lambda h: datetime.timedelta(hours=int(h)),
            help="time step in hours between times.",
        )
        parser.add_argument(
            "--end-time",
            default="2018-12-01",
            help="final time (inclusive).",
            type=datetime.datetime.fromisoformat,
        )

    @staticmethod
    def from_args(args) -> List[datetime.datetime]:
        """parse the command line arguments and return a TimeRange"""
        return get_times(args.start_time, args.end_time, args.time_step)


def get_times(start_time: datetime, end_time: datetime, step: datetime.timedelta):
    # the IFS data Jaideep downloaded only has 668 steps (up to end of november 2018)
    times = []
    time = start_time
    while time <= end_time:
        times.append(time)
        time += step
    return times


def add_model_args(parser: argparse.ArgumentParser, required=False):
    if required:
        parser.add_argument("model", type=str)
    else:
        parser.add_argument("--model", type=str)
    parser.add_argument(
        "--model-metadata",
        type=str,
        help="metadata.json file. Defaults to the metadata.json in the model package.",
        default="",
    )


def model_from_args(args, device):
    if args.model_metadata:
        with open(args.model_metadata) as f:
            metadata = schema.Model.parse_raw(f.read())
    else:
        metadata = None

    return networks.get_model(args.model, device=device, metadata=metadata)
