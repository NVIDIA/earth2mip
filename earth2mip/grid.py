# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import dataclasses
from typing import List

import numpy as np

from earth2mip import schema


@dataclasses.dataclass(frozen=True)
class LatLonGrid:

    lat: List[float]
    lon: List[float]

    @property
    def shape(self):
        return (len(self.lat), len(self.lon))


def equiangular_lat_lon_grid(
    nlat: int, nlon: int, includes_south_pole: bool = True
) -> LatLonGrid:
    """A regular lat-lon grid

    Lat is ordered from 90 to -90. Includes -90 and only if if includes_south_pole is True.
    Lon is ordered from 0 to 360. includes 0, but not 360.

    """  # noqa
    lat = np.linspace(90, -90, nlat, endpoint=includes_south_pole)
    lon = np.linspace(0, 360, nlon, endpoint=False)
    return LatLonGrid(lat.tolist(), lon.tolist())


def from_enum(grid_enum: schema.Grid) -> LatLonGrid:
    if grid_enum == schema.Grid.grid_720x1440:
        return equiangular_lat_lon_grid(720, 1440, includes_south_pole=False)
    elif grid_enum == schema.Grid.grid_721x1440:
        return equiangular_lat_lon_grid(721, 1440)
    elif grid_enum == schema.Grid.s2s_challenge:
        return equiangular_lat_lon_grid(181, 360)
    else:
        raise ValueError(f"Unknown grid {grid_enum}")
