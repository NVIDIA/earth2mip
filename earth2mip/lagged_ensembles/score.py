# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import logging

import torch

import earth2mip.grid
from earth2mip.crps import crps_from_empirical_cdf

logger = logging.getLogger(__name__)


def weighted_average(x, w, dim):
    return torch.mean(x * w, dim) / torch.mean(w, dim)


def area_average(grid: earth2mip.grid.LatLonGrid, x: torch.Tensor):
    lat = torch.tensor(grid.lat, device=x.device)[:, None]
    cos_lat = torch.cos(torch.deg2rad(lat))
    return weighted_average(x, cos_lat, dim=[-2, -1])


def score(
    grid: earth2mip.grid.LatLonGrid,
    ensemble: dict[int, torch.Tensor],
    obs: torch.Tensor,
) -> dict[str, torch.Tensor]:
    """Set of standardized scores for lagged ensembles

    Includes:
    - crps
    - mse of ensemble mean
    - variance about ensemble mean
    - MSE of the first ensemble member

    Args:
        ensemble: mapping of lag to (c, ...)
        obs: (c, ...)

    Returns:
        dict of str to (channel,) shaped metrics tensors.

    """
    # need to run this after since pandas.Index doesn't support cupy
    out = {}
    ens = torch.stack(list(ensemble.values()), dim=0)
    ensemble_dim = 0

    # compute all the metrics
    num = crps_from_empirical_cdf(obs, ens)
    out["crps"] = area_average(grid, num)

    num = (ens.mean(ensemble_dim) - obs) ** 2
    out["MSE_mean"] = area_average(grid, num)

    num = ens.var(ensemble_dim, unbiased=True)
    out["variance"] = area_average(grid, num)

    if 0 in ensemble:
        i = list(ensemble.keys()).index(0)
        num = (ens[i] - obs) ** 2
        out["MSE_det"] = area_average(grid, num)

    return {k: v.ravel() for k, v in out.items()}
