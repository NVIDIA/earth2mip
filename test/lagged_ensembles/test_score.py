# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import numpy as np
import pytest
import torch

from earth2mip.grid import equiangular_lat_lon_grid
from earth2mip.lagged_ensembles.score import score


def _run_score(nlat, nlon, nchannel, device="cpu"):
    # Create test data
    grid = equiangular_lat_lon_grid(nlat, nlon)

    ngrid = np.prod(grid.shape)
    ntot = ngrid * nchannel
    arr = torch.arange(ntot).reshape(1, nchannel, *grid.shape) / ntot
    arr = arr.to(device)
    ensemble = {
        0: arr - 1,
        1: arr,
        -1: arr + 1,
    }
    obs = torch.zeros_like(ensemble[0][0]).to(device)

    # Call the score function
    return score(grid, ensemble, obs)


def test_score_regression(regtest):
    # Create test data
    result = _run_score(12, 14, 10)
    # Assert the output shape
    for key in sorted(result):
        print(key, file=regtest)
        with np.printoptions(precision=3):
            print(result[key].numpy(), file=regtest)


@pytest.mark.slow
def test_score_large_data():
    """Make a test to avoid ooms"""
    torch.cuda.reset_peak_memory_stats()
    _run_score(721, 1440, 200, device="cuda")
    print(torch.cuda.memory_summary())
