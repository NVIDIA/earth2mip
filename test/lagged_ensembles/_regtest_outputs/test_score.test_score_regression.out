MSE_det
[0.903 0.723 0.563 0.423 0.303 0.203 0.123 0.063 0.023 0.003]
MSE_mean
[0.003 0.023 0.063 0.123 0.203 0.303 0.423 0.562 0.722 0.902]
crps
[0.239 0.272 0.305 0.339 0.372 0.405 0.439 0.472 0.505 0.539]
variance
[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
